README.md
# compose-webservices

Multiple webservices using `docker-compose`, behind `nginx-proxy`

## Create `.env` file

Create a file to store your environment variables

```shell
# Email for letsencrypt cert renew
EMAIL=me@localhost.com
# You server's domain name
DOMAIN=localhost.com
```

## Start services

At first you need to start the `nginx` service, it includes `nginx`, `letsencrypt` and `whoami` containers

```shell
docker-compose --project-name nginx up -d --force-recreate --remove-orphans
```

### Code Server

This is [VS Code](https://github.com/Microsoft/vscode) running on a remote server, accessible through the browser.

```shell
docker-compose --project-name code -f docker-code.yml up -d --build --force-recreate --remove-orphans
```

### Gitlab

```shell
docker-compose --project-name gitlab -f docker-gitlab.yml up -d --force-recreate --remove-orphans
```

### Jenkins

```shell
docker-compose --project-name jenkins -f docker-jenkins.yml up -d --force-recreate --remove-orphans
```

### Netdata

```shell
docker-compose --project-name netdata -f docker-netdata.yml up -d --force-recreate --remove-orphans
```

### Portainer

```shell
docker-compose --project-name portainer -f docker-portainer.yml up -d --force-recreate --remove-orphans
```

Generate new htpasswd, edit `command` property on `docker-portainer.yml`

```shell
docker run --rm httpd:2.4-alpine htpasswd -nbB admin <password> | cut -d ":" -f 2
```

* If any password passed though cli has an `$` you must escape it using double `$$`

### Node

A simple docker-compose example combining Nginx, Node and Mongo, this works by itself and shouldn't run with the other nginx instance. This part was obtained from this [post](https://medium.com/faun/learn-docker-in-5-days-day-5-docker-compose-11af7b9298db).

```shell
docker-compose --project-name node -f docker-node.yml up -d --force-recreate --remove-orphans
```

* See nginx running at: [localhost](http://localhost.com)
* See node app running at: [localhost/node](http://localhost.com/node)

## Test if it works locally

Use **cURL** it works you'll see the something like **I'm 999999999**:

```shell
curl -H "Host: whoami.localhost.com" localhost
```

If you have **Windows** or **macOS** add the following on:

* macOS: `/etc/hosts`
* Windows: `C:\Windows\System32\drivers\etc\hosts`

```shell
127.0.0.1 localhost.com
127.0.0.1 code.localhost.com
127.0.0.1 whoami.localhost.com
127.0.0.1 gitlab.localhost.com
127.0.0.1 jenkins.localhost.com
127.0.0.1 netdata.localhost.com
127.0.0.1 portainer.localhost.com
127.0.0.1 wordpress.localhost.com
127.0.0.1 db.wordpress.localhost.com
```

Then visit [whoami.localhost.com](http://whoami.localhost.com)` with your favorite browser

## Digital Ocean

[Create a Droplet](https://m.do.co/c/856dc39cd657)

## Watch large workspace VSCode

* Follow this [guide](https://code.visualstudio.com/docs/setup/linux#_visual-studio-code-is-unable-to-watch-for-file-changes-in-this-large-workspace-error-enospc)


CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Released]

## [1.0.0] - 2019-10-08
### Added
- VSCode settings `.vscode/settings.json`
- CI deploy with `.gitlab-ci.yml`
- Build with `docker-compose.yml` and `Dockerfile`
- New `CHANGELOG`, `CONTRIBUTING.md` and `LICENSE`
### Changed
- This project has nginx proxy auth
### Removed
- .DS_Store
- .directory
- desktop.ini


node/server.js
const compression = require('compression')
const mongoose = require('mongoose')
const express = require('express')
const helmet = require('helmet')
const { join } = require('path')
const cors = require('cors')
const app = express()

app.use(helmet()) // Security for browser
app.disable('etag') // Disable 304 not modified http status
app.use(compression()) // Compress responses
app.set('json spaces', 0) // # of spaces to indent json
app.disable('x-powered-by') // Avoid 0-day exploits for express
app.set('trust proxy', true) // sets req.ip
app.use(cors({ origin: '*' })) // CORS everywhere allowed
app.set('strict routing', false) // makes /foo and /foo/ the same
app.set('case sensitive routing', false) // makes /foo and /Foo the same

var PORT = 3000
var DB_NAME = 'nginx'
var DATABASE = `mongodb://localhost:27017/${DB_NAME}`
if (process.env.PRODUCTION) DATABASE = `mongodb://mongo/${DB_NAME}`

app.get('/*', function (req, res) { // Serving app with Client-Side Routing
  res.sendFile(join(__dirname, 'index.html')) // Serve file index.html
})

// https://mongoosejs.com/docs/deprecations.html
mongoose.set('useUnifiedTopology', true)
mongoose.set('useFindAndModify', false)
mongoose.set('useNewUrlParser', true)
mongoose.set('useCreateIndex', true)

mongoose.connect(DATABASE)
  .then(() => {
    console.log(`Database connected to ${DATABASE}`)
  })
  .then(() => app.listen(PORT, function () {
    console.log(`App connected on port ${PORT}`)
    app.timeout = 1000 * 10 // 10 secs
    app.emit('app.started')
  }))
  .catch(error => {
    console.error('Error at server startup', error)
  })


index.js
const compression = require('compression')
const mongoose = require('mongoose')
const { existsSync } = require('fs')
const express = require('express')
const helmet = require('helmet')
const { join } = require('path')
const cors = require('cors')
const app = express()

app.use(helmet()) // Security for browser
app.disable('etag') // Disable 304 not modified http status
app.use(compression()) // Compress responses
app.set('json spaces', 0) // # of spaces to indent json
app.disable('x-powered-by') // Avoid 0-day exploits for express
app.set('trust proxy', true) // sets req.ip
app.use(cors({ origin: '*' })) // CORS everywhere allowed
app.set('strict routing', false) // makes /foo and /foo/ the same
app.set('case sensitive routing', false) // makes /foo and /Foo the same

var PORT = 3000
var DB_NAME = 'nginx'
var DATABASE = `mongodb://localhost:27017/${DB_NAME}`
if (process.env.PRODUCTION) DATABASE = `mongodb://mongo/${DB_NAME}`

try {
  if (existsSync('./build')) { // If build folder exists
    app.use(express.static(join(__dirname, 'build'))) // Serve static files
    app.get('/*', function (req, res) { // Serving app with Client-Side Routing
      res.sendFile(join(__dirname, 'build', 'index.html')) // Serve file index.html
    })
  } else {
    app.get('/*', (req, res) => {
      res.status(404).json({ message: 'The front has not been build' })
    })
  }
} catch (error) {
  console.error(error)
}

// https://mongoosejs.com/docs/deprecations.html
mongoose.set('useUnifiedTopology', true)
mongoose.set('useFindAndModify', false)
mongoose.set('useNewUrlParser', true)
mongoose.set('useCreateIndex', true)

mongoose.connect(DATABASE)
  .then(() => {
    console.log(`Database connected to ${DATABASE}`)
  })
  .then(() => app.listen(PORT, function () {
    console.log(`App connected on port ${PORT}`)
    app.timeout = 1000 * 10 // 10 secs
    app.emit('app.started')
  }))
  .catch(error => {
    console.error('Error at server startup', error)
  })
